File: apps/commerce-api/src/main/java/com/loopers/domain/inventory/InventoryService.java
package com.loopers.domain.inventory;

import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;

import lombok.RequiredArgsConstructor;

@Component
@RequiredArgsConstructor
public class InventoryService {
    private final InventoryRepository inventoryRepository;

    @Transactional
    public void reverse(InventoryCommand.Reverse command) {
        Inventories inventories = new Inventories(inventoryRepository.findAll(command.productIds(), command.productOptionIds()));
        inventories.reverse(command);
    }

    @Transactional(readOnly = true)
    public Map<Long, Integer> getStockQuantities(List<Long> optionIds) {
        return inventoryRepository.findAll(optionIds)
            .stream()
            .collect(Collectors.toMap(
                Inventory::getProductOptionId,
                Inventory::getQuantity
            ));
    }
}

File: apps/commerce-streamer/src/main/kotlin/com/loopers/domain/inventory/InventoryService.kt
package com.loopers.domain.inventory

import com.loopers.support.error.CoreException
import com.loopers.support.error.ErrorType
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional

@Component
class InventoryService(
    private val inventoryRepository: InventoryRepository
) {
    @Transactional
    fun deduct(command: InventoryCommand.Deduct) {
        val inventories = inventoryRepository.findAll(command.productIds(), command.productOptionIds())

        command.options.forEach { option ->
            val inventory = inventories.find {
                it.productId == option.productId && it.productOptionId == option.productOptionId
            } ?: throw CoreException(ErrorType.NOT_FOUND)
            inventory.deduct(option.quantity)
        }
    }
}
